File: design browser.java
class Node {
    String data;
    Node next;
    Node back;

    // Default constructor
    Node() {
        this.data = "";
        this.next = null;
        this.back = null;
    }

    // Constructor with one parameter
    Node(String data) {
        this.data = data;
        this.next = null;
        this.back = null;
    }

    // Constructor with multiple parameters
    Node(String data, Node next, Node back) {
        this.data = data;
        this.next = next;
        this.back = back;
    }
}

class BrowserHistory {
    Node currentPage;

    // Constructor for BrowserHistory
    BrowserHistory(String homepage) {
        currentPage = new Node(homepage);
    }

    // Method to visit a new URL
    public void visit(String url) {
        Node newNode = new Node(url);
        currentPage.next = newNode;
        newNode.back = currentPage;
        currentPage = newNode;
    }

    // Method to go back in history
    public String back(int steps) {
        while (steps > 0 && currentPage.back != null) {
            currentPage = currentPage.back;
            steps--;
        }
        return currentPage.data;
    }

    // Method to go forward in history
    public String forward(int steps) {
        while (steps > 0 && currentPage.next != null) {
            currentPage = currentPage.next;
            steps--;
        }
        return currentPage.data;
    }
}

File: design browser.cpp
class Node{
    public:
    string data;
    Node* next;
    Node* back;
    Node():data(""),next(nullptr),back(nullptr){};
    Node(string x):data(x),next(nullptr),back(nullptr){}
    Node(string x, Node *next,Node *random):data(x),next(next),back(random){}
};

class BrowserHistory {
public:
    Node* currentpage;
    BrowserHistory(string homepage) {
        currentpage=new Node(homepage);
    }
    
    void visit(string url) {
        Node * newNode =new Node(url);
        currentpage->next=newNode;
        newNode->back=currentpage;
        currentpage=newNode;
    }
    
    string back(int steps) {
       while(steps){
        if(currentpage->back){
            currentpage=currentpage->back;
        }else break;
        steps--;
       } 
       return currentpage->data;
    }
    
    string forward(int steps) {
        while(steps){
            if(currentpage->next)
            currentpage=currentpage->next;
            else break;
            steps--;
        }
        return currentpage->data;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */